compiler/rustc_data_structures/src/graph/scc/mod.rs

```diff
@@ -392,74 +401,186 @@ where
+
+                    debug_assert!(matches!(self.node_states[node], NodeState::NotVisited));
+
+                    // Push `node` onto the stack.
+                    self.node_states[node] = NodeState::BeingVisited { depth };
+                    self.node_stack.push(node);
+
+                    // Walk each successor of the node, looking to see if any of
+                    // them can reach a node that is presently on the stack. If
+                    // so, that means they can also reach us.
+                    *successors_len = self.successors_stack.len();
+                    // Set and return a reference, this is currently empty.
+                    iter.get_or_insert(self.graph.successors(node))
+                }
+            };
+
+            // Now that iter is initialized, this is fixed.
+            let successors_len = *successors_len;
+
+            // Handle return value.
+            // FIXME : this should be handled without duplicating the matching of `walk` in the same
+            // manner as in the loop following, going over the next successors. However, to achieve
+            // this we would have the chain the iterators without borrowing `self` since that would
+            // bar us from pushing the `successor_scc_index`, which requires access to `self`.
+            if let Some(walk) = return_value.take() {
+                match walk {
+                    WalkReturn::Cycle { min_depth: successor_min_depth } => {
+                        // Track the minimum depth we can reach.
+                        assert!(successor_min_depth <= depth);
+                        if successor_min_depth < *min_depth {
+                            debug!(
+                                "walk_unvisited_node: node = {:?} successor_min_depth = {:?}",
+                                node, successor_min_depth
+                            );
+                            *min_depth = successor_min_depth;
+                            *min_cycle_root = *successor_node;
+                        }
+                    }
+
+                    WalkReturn::Complete { scc_index: successor_scc_index } => {
+                        // Push the completed SCC indices onto
+                        // the `successors_stack` for later.
                         debug!(
-                            "walk_unvisited_node: node = {:?} successor_min_depth = {:?}",
-                            node, successor_min_depth
+                            "walk_unvisited_node: node = {:?} successor_scc_index = {:?}",
+                            node, successor_scc_index
                         );
-                        min_depth = successor_min_depth;
-                        min_cycle_root = successor_node;
+                        self.successors_stack.push(successor_scc_index);
                     }
                 }
+            }
+
+            for successor_node in successors.by_ref() {
```

Comment:
Something about this code feels more complex than I expect. I think it's the duplication you mention above. I guess I expect us to process exactly one successor for each iteration of the outer loop. This would I suppose be less efficient overall but simpler, I think, to read?

Reply:
I've changed it to use `Iterator::chain` such that the same `match` block is used.